modules/tinymce/src/core/main/ts/api/dom/DOMUtils.ts

```diff
@@ -1090,43 +1089,45 @@ function DOMUtils(doc: Document, settings: Partial<DOMUtilsSettings> = {}): DOMU
     }
   };
 
-  const bind = <T = any>(target: Target, name: string, func: EventUtilsCallback<T>, scope?: any): EventUtilsCallback<T>[] | EventUtilsCallback<T> => {
-    if (Tools.isArray(target)) {
+  const bind = <
+    K extends string,
+    T extends Target,
+    F extends EventUtilsCallback<MappedEvent<HTMLElementEventMap, K>>
+  >(target: T, name: string, func: F, scope?: any): T extends [] ? F[] : F => {
+    if (Tools.isArray<Node | Window>(target)) {
       let i = target.length;
-      const rv = [];
+      const rv = [] as F[];
 
       while (i--) {
         rv[i] = bind(target[i], name, func, scope);
       }
 
-      return rv;
+      return rv as T extends [] ? F[] : F;
```

Comment:
This really shouldn't be needed... I'm going to have to look into this more.

Reply:
I wish it wasn't, but I'm not sure if there's a way around it. We give the type checker the following information:

```ts
target: T; // from the declaration
target is Array<Node | Window>; // from Tools.isArray
```

And yet, it is unable to infer:

```ts
T: Array<Node | Window>;
```

Because as far as typescript is concerned, the `target: T` translates to "target is **at least** T" (thanks implicit subtyping rules), meaning that `target: Array<Node | Window>` is interpreted as `target T & Array<Node | Window>` rather than `target: T; T is Array<Node | Window>`.

As such, when we return `F[]` (`U[]` in future) it complains that we don't actually know that `T extends []`